source/agora/consensus/state/UTXOSet.d

```diff
@@ -156,41 +234,98 @@ public class UTXOSet
 
     ***************************************************************************/
 
-    public UTXO[Hash] getUTXOs (const ref PublicKey pubkey) @safe nothrow
+    public UTXO[Hash] getUTXOs (const ref PublicKey pubkey) nothrow @safe
     {
         return this.utxo_db.getUTXOs(pubkey);
     }
 
-    /***************************************************************************
+    ///
+    public override bool peekUTXO (Hash utxo, out UTXO value) nothrow @safe
+    {
+        return this.utxo_db.find(utxo, value);
+    }
 
-        Find an UTXO in the UTXO set.
+    ///
+    protected override void remove (Hash utxo) @safe
+    {
+        this.utxo_db.remove(utxo);
+    }
 
-        Params:
-            hash = the hash of the UTXO (`hashMulti(tx_hash, index)`)
-            output = will contain the UTXO if found
+    ///
+    protected override void add (Hash utxo, UTXO value) @safe
+    {
+        this.utxo_db[utxo] = value;
+    }
+}
 
-        Return:
-            Return true if the UTXO was found
+/*******************************************************************************
 
-    ***************************************************************************/
+    This is a simple UTXOSet, used when the AA behavior is desired
+
+    Most unittests do not need a fully-fledged UTXOSet with all the DB and
+    serialization that comes with it, instead relying on an associative array
+    and a delegate.
+
+    Since this pattern is so common, this class is offered as a mean to achieve
+    this without code duplication. See issue #501 for history.
+
+    Note that this should *NOT* be used to replace the above UTXOSet,
+    when for example doing integration tests with LocalRest.
+
+*******************************************************************************/
 
-    private bool findUTXO (Hash utxo, out UTXO value)
-        nothrow @safe
+public class TestUTXOSet : UTXOCache
+{
+    /// UTXO cache backed by an AA
+    public UTXO[Hash] storage;
+
+    ///
+    alias storage this;
+
+    /// Short hand to add a transaction
+    public void put (const Transaction tx) nothrow @safe
     {
-        if (utxo in this.used_utxos)
+        Hash txhash = hashFull(tx);
+        foreach (size_t idx, ref output_; tx.outputs)
         {
-            log.trace("findUTXO: utxo_hash {} found in used_utxos: {}", utxo, used_utxos);
-            return false;  // double-spend
+            Hash h = UTXO.getHash(txhash, idx);
+            UTXO v = {
+                type: tx.type,
+                output: output_
+            };
+            this.storage[h] = v;
         }
+    }
+
+    /// Workaround 20559...
+    public void clear ()
+    {
+        this.storage.clear();
+    }
```

Comment:
Why is this a workaround ?

Reply:
https://issues.dlang.org/show_bug.cgi?id=20559

But I fixed it and it's not needed since v2.090.1.

Reply:
Oh ok. I'll remove this comment then.